examples/fluids/src/smartsim/sgs_dd_training.c

```diff
@@ -244,71 +267,84 @@ PetscErrorCode TSMonitor_SGS_DD_Training(TS ts, PetscInt step_num, PetscReal sol
   if (step_num % sgs_dd_train->write_data_interval != 0) PetscFunctionReturn(PETSC_SUCCESS);
   PetscCall(DMGetGlobalVector(sgs_dd_train->dm_dd_training, &TrainingData));
 
-  PetscCall(PetscLogEventBegin(FLUIDS_TrainDataCompute, 0, 0, 0, 0));
-  {  // -- Compute and assemble training data
-    Vec          FilteredVelocityGradient, FilteredFields, FilteredFields_loc;
-    PetscMemType filtered_fields_mem_type;
-    CeedVector   filtered_fields;
+  for (PetscInt filter_index = 0; filter_index < sgs_dd_train->num_filter_widths; filter_index++) {
+    PetscCall(PetscLogEventBegin(FLUIDS_TrainDataCompute, 0, 0, 0, 0));
+    {  // -- Compute and assemble training data
+      Vec          FilteredVelocityGradient, FilteredFields, FilteredFields_loc;
+      PetscMemType filtered_fields_mem_type;
+      CeedVector   filtered_fields;
 
-    PetscCall(DMGetGlobalVector(user->diff_filter->dm_filter, &FilteredFields));
-    PetscCall(DMGetLocalVector(user->diff_filter->dm_filter, &FilteredFields_loc));
+      {  // Set filter width for the current solve
+        double       filter_width_scaling[3];
+        CeedOperator op_mat;
+        Mat          A_mat;
 
-    PetscCall(DifferentialFilterApply(user, solution_time, Q, FilteredFields));
-    PetscCall(DMGlobalToLocal(user->diff_filter->dm_filter, FilteredFields, INSERT_VALUES, FilteredFields_loc));
+        for (int j = 0; j < 3; j++) filter_width_scaling[j] = sgs_dd_train->filter_widths[filter_index];
+        PetscCall(KSPGetOperators(user->diff_filter->ksp, &A_mat, NULL));
+        PetscCall(MatCeedGetCeedOperators(A_mat, &op_mat, NULL));
+        PetscCall(CeedOperatorSetContextDouble(op_mat, user->diff_filter->filter_width_scaling_label, filter_width_scaling));
+      }
 
-    PetscCall(DMGetGlobalVector(sgs_dd_train->filtered_grad_velo_proj->dm, &FilteredVelocityGradient));
-    PetscCall(VelocityGradientProjectionApply(sgs_dd_train->filtered_grad_velo_proj, FilteredFields_loc, FilteredVelocityGradient));
+      PetscCall(DMGetGlobalVector(user->diff_filter->dm_filter, &FilteredFields));
+      PetscCall(DMGetLocalVector(user->diff_filter->dm_filter, &FilteredFields_loc));
 
-    {
-      CeedOperatorField op_field;
-      PetscCallCeed(ceed, CeedOperatorGetFieldByName(sgs_dd_train->op_training_data_calc_ctx->op, "q", &op_field));
-      PetscCallCeed(ceed, CeedOperatorFieldGetVector(op_field, &filtered_fields));
-    }
-    PetscCall(VecPetscToCeed(FilteredFields_loc, &filtered_fields_mem_type, filtered_fields));  // filtered_fields is an implicit input
-
-    PetscCall(ApplyCeedOperatorGlobalToGlobal(FilteredVelocityGradient, TrainingData, sgs_dd_train->op_training_data_calc_ctx));
+      PetscCall(DifferentialFilterApply(user, solution_time, Q, FilteredFields));
```

Comment:
Note we are calling `DifferentialFilterApply` multiple times per filter width. This is slightly inefficient, as the RHS gets recalculated each time, even though the value is the same.

One way around this would be to save the RHS result and track the `PetscObjectState` of the solution vector used to create the RHS vector; if the solution vector state hasn't changed, then the previous existing RHS doesn't need to be updated. However, [`PetscObjectStateGet`](https://petsc.org/release/manualpages/Sys/PetscObjectStateGet/#petscobjectstateget) is kept behind `include/petsc/private/petscimpl.h`. 

@jedbrown @jeremylt IDK what y'all's thoughts are behind using functions in `include/petsc/private/petscimpl.h`, but I'm guessing I probably shouldn't be doing that?

Reply:
I'm using it inside of `ratel-pmg.c`. I wonder if maybe `VecGetState` and `MatGetState` should be public PETSc functions?

Reply:
Ok, I've implemented the `PetscObjectState` solution in https://github.com/CEED/libCEED/pull/1558/commits/56f5e8119536c76269613d7f3474b8135eed8bb0



> I wonder if maybe `VecGetState` and `MatGetState` should be public PETSc functions?

Not sure. Probably though? I don't see any reason for them not to be.

Reply:
https://gitlab.com/petsc/petsc/-/merge_requests/7470